File: Algorithm/String Matching/Trie树的构建与查询.py
class TrieNode:
    data = None
    isEndingChar = False
    children = [None for i in range(0,26)]
    def __init__(self,data):
        self.data = data

    def __repr__(self):
        return str(self.data)

class Trie:
    def __init__(self):
        self.root = TrieNode('/')

    # 往 Trie 树中插入一个字符串
    def insert(self,text):
        p = self.root
        for i in range(0,len(text)):
            index = ord(text[i]) - ord('a')
            if (p.children[index] == None):
                newNode = TrieNode(text[i])
                p.children[index] = newNode
            p = p.children[index]
        p.isEndingChar = True

    # 在 Trie 树中查找一个字符串
    def find(self,pattern):
        p = self.root;
        for i in range(0,len(pattern)):
            index = ord(pattern[i]) - ord('a')
            if p.children[index] == None: 
                return False        #不存在 pattern
            p = p.children[index]
        if p.isEndingChar == False:
            return False            #不能完全匹配，只是前缀
        else:
            return True             #找到 pattern


trie1 = Trie()
trie1.insert("abc")
trie1.insert("def")
trie1.insert("hig")

result = trie1.find("abc")

if result:
    print("已查找到模式串")
else:
    print("未查找到模式串")
  

File: Algorithm/String Matching/Trie树的构建与查询.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <malloc.h>

struct TrieNode{
    char data;
    struct TrieNode *children[26];
    int isEndingChar;
};

// 往 Trie 树中插入一个字符串
void insert(struct TrieNode *root,char* text) {

    struct TrieNode *p = root;
    for (int i = 0; i < strlen(text); ++i) {
      int index = text[i] - 'a';
      if (p->children[index] == NULL) {
        struct TrieNode *newNode=(struct TrieNode *)calloc(1,sizeof(struct TrieNode));
        newNode->data = text[i];
        p->children[index] = newNode;
      }
      p = p->children[index];
      p->isEndingChar = 0;
    }
    p->isEndingChar = 1;
    printf("字符串%s已经成功插入Trie树\n",text);
}

// 在 Trie 树中查找一个字符串
int find(struct TrieNode *root,char* pattern) {
    struct TrieNode *p = root;
    for (int i = 0; i < strlen(pattern); ++i) {
      int index = pattern[i] - 'a';
      if (p->children[index] == NULL) {
        return 0; // 不存在 pattern
      }
      p = p->children[index];
    }
    if (p->isEndingChar == 0) return 0; // 不能完全匹配，只是前缀
    else return 1; // 找到 pattern
}

int main(){
    struct TrieNode* root = calloc(1,sizeof(struct TrieNode));
    root->data ='/';
    root->isEndingChar =0;

    insert(root,"abc");
    insert(root,"def");
    insert(root,"ghi");
    int result = find(root,"abc");

    if(result){
        printf("已查找到模式串");
    }else{
        printf("未查找到模式串");
    }
}

